api/src/server.rs

```diff
@@ -0,0 +1,388 @@
+
+use ceramic_api_server::server::MakeService;
+use ceramic_api_server::Api;
+use std::error::Error;
+use swagger::ApiError;
+
+#[async_trait]
+impl<C, R> Api<C> for Server<C, R>
+where
+    C: Send + Sync,
+    R: Recon + Sync,
+{
+    async fn ceramic_events_post(
+        &self,
+        event: models::Event,
+        _context: &C,
+    ) -> Result<CeramicEventsPostResponse, ApiError> {
+        debug!(event_id = event.event_id, "ceramic_events_post");
+        let event_id = decode_event_id(&event.event_id)?;
+        self.recon.insert_key(&event_id);
+        Ok(CeramicEventsPostResponse::Success)
+    }
+
+    async fn ceramic_subscribe_sort_value_get(
+        &self,
+        sort_value: String,
+        controller: Option<String>,
+        stream_id: Option<String>,
+        _context: &C,
+    ) -> Result<CeramicSubscribeSortValueGetResponse, ApiError> {
+        // Construct start and stop event id based on provided data.
+        let start_builder = EventId::builder()
+            .with_network(&self.network)
+            .with_sort_value(&sort_value);
+        let stop_builder = EventId::builder()
+            .with_network(&self.network)
+            .with_sort_value(&sort_value);
+        let (start_builder, stop_builder) = if let Some(controller) = controller {
+            if let Some(stream_id) = stream_id {
+                let stream_id = StreamId::from_str(&stream_id)
+                    .map_err(|err| ApiError(format!("stream_id: {}", err)))?;
+                // We have both controller and stream id
+                (
+                    start_builder
+                        .with_controller(&controller)
+                        .with_init(&stream_id.cid),
+                    stop_builder
+                        .with_controller(&controller)
+                        .with_init(&stream_id.cid),
+                )
+            } else {
+                // We have a controller without a stream id
+                (
+                    start_builder.with_controller(&controller).with_min_init(),
+                    stop_builder.with_controller(&controller).with_max_init(),
+                )
+            }
+        } else {
+            if stream_id.is_some() {
+                return Err(ApiError(
+                    "controller is required if stream_id is specified".to_owned(),
+                ));
+            }
+            // We have no controller or stream id
+            (
+                start_builder.with_min_controller().with_min_init(),
+                stop_builder.with_max_controller().with_max_init(),
+            )
+        };
+        let start = start_builder.with_min_event_height().build_fencepost();
+        let stop = stop_builder.with_max_event_height().build_fencepost();
+
+        debug!(%start, %stop, "subscribe");
+        Ok(CeramicSubscribeSortValueGetResponse::Success(
+            self.recon
+                .range((Bound::Included(start), Bound::Excluded(stop)))
+                .into_iter()
```

Comment:
What order will this sort in?

Reply:
Ascending lexical order of EventId

Reply:
So this api currently returns all known eventIds in the requested range?

Reply:
yes, that is the intent. Should some event Ids be skipped?

Reply:
So how would this replace pubsub then? Don't we need the offset logic here?

Reply:
The way I imagined this working is:
1. js-ceramic subscribes to the range
2. rust-ceramic returns all known ranges
3. js-ceramic subscribes again to get new updates

We talked about 3 using long polling to be able to get new events as they come in. Am I missing something about how this as implemented currently would be used?